app/worker.py

```diff
@@ -70,11 +70,14 @@ def handler(event, context):
             current_components[cid]['status'] = data['status']
             changed.append(current_components[cid]['name'])
 
-    # If components stop reporting set their status to warning.
+    # If components stop reporting set their status to warning if not already
+    # set.
     for cid in current_components.keys():
         if cid not in updated_components:
-            current_components[cid]['status'] = config.STATUS_MAP['warning']['name']
-            changed.append(current_components[cid]['name'])
+            warning_status = config.STATUS_MAP['warning']['name']
+            if current_components[cid]['status'] != warning_status:
```

Comment:
Is the `status` key guaranteed to exist in `current_components[cid]` at this point? If not we should guard against that possibility with a `.get('status', '')` here instead of direct dict access to avoid the possible exception.